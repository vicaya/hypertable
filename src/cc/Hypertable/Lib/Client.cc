/** -*- c++ -*-
 * Copyright (C) 2008 Doug Judd (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 2 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include "Common/Compat.h"
#include <cassert>
#include <cstdlib>

extern "C" {
#include <poll.h>
}

#include <boost/algorithm/string.hpp>

#include "AsyncComm/Comm.h"
#include "AsyncComm/ReactorFactory.h"

#include "Common/Init.h"
#include "Common/Error.h"
#include "Common/InetAddr.h"
#include "Common/Logger.h"
#include "Common/ScopeGuard.h"
#include "Common/System.h"
#include "Common/Timer.h"

#include "Hyperspace/DirEntry.h"
#include "Hypertable/Lib/Config.h"

#include "Client.h"
#include "HqlCommandInterpreter.h"

using namespace std;
using namespace Hypertable;
using namespace Hyperspace;
using namespace Config;

Client::Client(const String &install_dir, const String &config_file,
               uint32_t default_timeout_ms)
  : m_timeout_ms(default_timeout_ms), m_install_dir(install_dir) {
  ScopedRecLock lock(rec_mutex);

  if (!properties)
    init_with_policy<DefaultCommPolicy>(0, 0);

  m_props = new Properties(config_file, file_desc());
  initialize();
}

Client::Client(const String &install_dir, uint32_t default_timeout_ms)
  : m_timeout_ms(default_timeout_ms), m_install_dir(install_dir) {
  ScopedRecLock lock(rec_mutex);

  if (!properties)
    init_with_policy<DefaultCommPolicy>(0, 0);

  if (m_install_dir.empty())
    m_install_dir = System::install_dir;

  m_props = properties;
  initialize();
}

void Client::create_namespace(const String &name, Namespace *base, bool create_intermediate) {

  String full_name;
  String sub_name = name;
  int flags=NameIdMapper::IS_NAMESPACE;

  if(create_intermediate)
    flags |= NameIdMapper::CREATE_INTERMEDIATE;

  if (base != NULL) {
    full_name = base->get_name() + '/';
  }
  full_name += sub_name;

  Namespace::canonicalize(&full_name);
  m_master_client->create_namespace(full_name, flags);
}

NamespacePtr Client::open_namespace(const String &name, Namespace *base) {
  String full_name;
  String sub_name = name;

  Namespace::canonicalize(&sub_name);

  if (base != NULL) {
    full_name = base->get_name() + '/';
  }

  full_name += sub_name;

  return m_namespace_cache->get(full_name);
}

bool Client::exists_namespace(const String &name, Namespace *base) {
  String id;
  bool is_namespace = false;
  String full_name;
  String sub_name = name;

  Namespace::canonicalize(&sub_name);

  if (base != NULL) {
    full_name = base->get_name() + '/';
  }

  full_name += sub_name;

  if (!m_namemap->name_to_id(full_name, id, &is_namespace) || !is_namespace)
    return false;

  // TODO: issue 11

  String namespace_file = m_toplevel_dir + "/tables/" + id;

  try {
    return m_hyperspace->exists(namespace_file);
  }
  catch(Exception &e) {
    HT_THROW(e.code(), e.what());
  }
}

void Client::drop_namespace(const String &name, Namespace *base, bool if_exists) {
  String full_name;
  String sub_name = name;

  Namespace::canonicalize(&sub_name);

  if (base != NULL) {
    full_name = base->get_name() + '/';
  }

  full_name += sub_name;

  m_namespace_cache->remove(full_name);
  m_master_client->drop_namespace(full_name, if_exists);
}

Hyperspace::SessionPtr& Client::get_hyperspace_session()
{
  return m_hyperspace;
}

void Client::close() {
  HT_WARN("close() is no longer supported");
}

void Client::shutdown() {
  m_master_client->shutdown();
}


HqlInterpreter *Client::create_hql_interpreter(bool immutable_namespace) {
  return new HqlInterpreter(this, this->m_conn_manager, immutable_namespace);
}

// ------------- PRIVATE METHODS -----------------
void Client::initialize() {
  uint32_t wait_time, remaining;
  uint32_t interval=5000;

  m_toplevel_dir = m_props->get_str("Hypertable.Directory");
  boost::trim_if(m_toplevel_dir, boost::is_any_of("/"));
  m_toplevel_dir = String("/") + m_toplevel_dir;

  m_comm = Comm::instance();
  m_conn_manager = new ConnectionManager(m_comm);

  if (m_timeout_ms == 0)
    m_timeout_ms = m_props->get_i32("Hypertable.Request.Timeout");

  m_hyperspace_reconnect = m_props->get_bool("Hyperspace.Session.Reconnect");

  m_hyperspace = new Hyperspace::Session(m_comm, m_props);

  m_namemap = new NameIdMapper(m_hyperspace, m_toplevel_dir);

  m_refresh_schema = m_props->get_bool("Hypertable.Client.RefreshSchema");

  Timer timer(m_timeout_ms, true);

  remaining = timer.remaining();
  wait_time = (remaining < interval) ? remaining : interval;

  while (!m_hyperspace->wait_for_connection(wait_time)) {

    if (timer.expired())
      HT_THROW_(Error::CONNECT_ERROR_HYPERSPACE);

    cout << "Waiting for connection to Hyperspace..." << endl;

    remaining = timer.remaining();
    wait_time = (remaining < interval) ? remaining : interval;
  }

  m_app_queue = new ApplicationQueue(m_props->
                                     get_i32("Hypertable.Client.Workers"));
  m_master_client = new MasterClient(m_conn_manager, m_hyperspace, m_toplevel_dir,
                                     m_timeout_ms, m_app_queue);

  m_master_client->initiate_connection();

  if (!m_master_client->wait_for_connection(timer))
    HT_THROW(Error::REQUEST_TIMEOUT, "Waiting for Master connection");

  m_range_locator = new RangeLocator(m_props, m_conn_manager, m_hyperspace,
                                     m_timeout_ms);
  m_table_cache = new  TableCache(m_props, m_range_locator, m_conn_manager, m_hyperspace,
                                  m_app_queue, m_namemap, m_timeout_ms);
  m_namespace_cache = new NamespaceCache(m_props, m_range_locator, m_conn_manager, m_hyperspace,
                                         m_app_queue, m_namemap, m_master_client,
                                         m_table_cache, m_timeout_ms);
}
