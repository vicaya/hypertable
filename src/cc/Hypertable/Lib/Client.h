/** -*- c++ -*-
 * Copyright (C) 2010 Doug Judd (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 2 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#ifndef HYPERTABLE_CLIENT_H
#define HYPERTABLE_CLIENT_H

#include "Common/Mutex.h"
#include "Common/ReferenceCount.h"
#include "Common/String.h"

#include "AsyncComm/ApplicationQueue.h"
#include "AsyncComm/ConnectionManager.h"
#include "Hyperspace/Session.h"

#include "MasterClient.h"
#include "NameIdMapper.h"
#include "NamespaceCache.h"
#include "TableCache.h"
#include "Namespace.h"

namespace Hypertable {

  class Comm;
  class HqlInterpreter;

  class Client : public ReferenceCount {
  public:

    /**
     * Constructs the object using the specified config file
     *
     * @param install_dir path to Hypertable installation directory
     * @param config_file name of configuration file
     * @param default_timeout_ms default method call timeout in milliseconds
     */
    Client(const String &install_dir, const String &config_file,
           uint32_t default_timeout_ms=0);

    /**
     * Constructs the object using the default config file
     *
     * @param install_dir path to Hypertable installation directory
     * @param default_timeout_ms default method call timeout in milliseconds
     */
    Client(const String &install_dir = String(), uint32_t default_timeout_ms=0);
    ~Client() {}

    /**
     * Creates a namespace
     *
     * @param name name of the namespace to create
     * @param base optional base Namespace (if specified the name parameter will be relative
     *        to base)
     * @param create_intermediate if true then create all non-existent intermediate namespaces
     */
    void create_namespace(const String &name, Namespace *base=NULL,
                          bool create_intermediate=false);

    /**
     * Opens a Namespace
     *
     * @param name name of the Namespace
     * @param base optional base Namespace (if specified the name parameter will be relative
     *        to base)
     * @return pointer to the Namespace object
     */
    NamespacePtr open_namespace(const String &name, Namespace *base=NULL);

    /**
     * Checks if the namespace exists
     *
     * @param name name of namespace
     * @param base optional base Namespace (if specified the name parameter will be relative
     *        to base)
     * @return true if namespace exists false ow
     */
    bool exists_namespace(const String &name, Namespace *base=NULL);

    /**
     * Removes a namespace.  This command instructs the Master to
     * remove a namespace from the system. The namespace must be empty, ie all tables and
     * namespaces under it must have been dropped already
     *
     * @param name namespace name
     * @param base optional base Namespace (if specified the name parameter will be relative
     *        to base)
     * @param if_exists don't throw an exception if table does not exist
     */
    void drop_namespace(const String &name, Namespace *base=NULL, bool if_exists=false);

    Hyperspace::SessionPtr& get_hyperspace_session();
    void close();
    void shutdown();

    /**
     *
     * @param immutable_namespace if true then the namespace can't be modified once its set
     */
    HqlInterpreter *create_hql_interpreter(bool immutable_namespace=true);

  private:

    void initialize();

    PropertiesPtr           m_props;
    Comm                   *m_comm;
    ConnectionManagerPtr    m_conn_manager;
    ApplicationQueuePtr     m_app_queue;
    Hyperspace::SessionPtr  m_hyperspace;
    NameIdMapperPtr         m_namemap;
    MasterClientPtr         m_master_client;
    RangeLocatorPtr         m_range_locator;
    uint32_t                m_timeout_ms;
    String                  m_install_dir;
    TableCachePtr           m_table_cache;
    NamespaceCachePtr       m_namespace_cache;
    Mutex                   m_mutex;
    bool                    m_hyperspace_reconnect;
    bool                    m_refresh_schema;
    String                  m_toplevel_dir;
  };

  typedef intrusive_ptr<Client> ClientPtr;

} // namespace Hypertable

#endif // HYPERTABLE_CLIENT_H
