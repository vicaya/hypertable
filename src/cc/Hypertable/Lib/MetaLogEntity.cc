/** -*- c++ -*-
 * Copyright (C) 2010 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 2 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */
#include "Common/Compat.h"
#include "Common/Checksum.h"
#include "Common/Error.h"
#include "Common/Logger.h"
#include "Common/Serialization.h"

#include "MetaLogEntity.h"

using namespace Hypertable;
using namespace Hypertable::MetaLog;

Entity::Entity(int32_t type) : header(type) { }

Entity::Entity(const EntityHeader &header_) : header(header_) { }


void Entity::encode_entry(uint8_t **bufp) {
  header.length = encoded_length();
  uint8_t *header_encode_position = *bufp;
  *bufp += EntityHeader::LENGTH;
  uint8_t *payload_encode_position = *bufp;
  encode(bufp);
  header.checksum = fletcher32(payload_encode_position, *bufp-payload_encode_position);
  header.encode(&header_encode_position);
  HT_ASSERT(header_encode_position == payload_encode_position);
}

void Entity::decode_entry(const uint8_t **bufp, size_t *remainp) {
  header.decode(bufp, remainp);
  const uint8_t *base = *bufp;
  decode(bufp, remainp);
  int32_t checksum = (int32_t)fletcher32(base, *bufp-base);
  if (checksum != header.checksum)
    HT_THROWF(Error::METALOG_CHECKSUM_MISMATCH,
              "MetaLog checksum mismatch header=%lx, computed=%lx",
              (Lu)header.checksum, (Lu)checksum);
}
