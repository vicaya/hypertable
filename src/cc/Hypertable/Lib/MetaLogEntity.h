/** -*- c++ -*-
 * Copyright (C) 2010 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 2 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#ifndef HYPERTABLE_METALOGENTITY_H
#define HYPERTABLE_METALOGENTITY_H

#include <iostream>

extern "C" {
#include <time.h>
}

#include <boost/algorithm/string.hpp>

#include "Common/ReferenceCount.h"

#include "MetaLogEntityHeader.h"

namespace Hypertable {

  namespace MetaLog {

    class Reader;
    class Writer;
    
    class Entity : public ReferenceCount {
    public:
      Entity(int32_t type);
      Entity(const EntityHeader &header_);
      virtual ~Entity() { }
      virtual size_t encoded_length() const { return 0; }
      virtual void encode(uint8_t **bufp) const { return; }
      virtual void decode(const uint8_t **bufp, size_t *remainp) { return; }

      virtual const String name() = 0;

      /**
       * Prints a textual representation of the entity state to the given ostream
       *
       * @param os ostream to print entity to
       */
      virtual void display(std::ostream &os) { }

      friend std::ostream &operator <<(std::ostream &os, Entity &entity);
      friend class Reader;
      friend class Writer;

    protected:
      void encode_entry(uint8_t **bufp);
      void decode_entry(const uint8_t **bufp, size_t *remainp);
      EntityHeader header;
    };
    typedef intrusive_ptr<Entity> EntityPtr;

    /**
     * ostream shift function for Entity objects
     */
    inline std::ostream &
    operator <<(std::ostream &os, Entity &entity) {
      os << "{MetaLog::Entity " << entity.name() << " header={";
      entity.header.display(os);
      os << "} payload={";
      entity.display(os);
      os << "}";
      return os;
    }

    namespace EntityType {
      enum {
        RECOVER = 0x00000001
      };
    }

    class EntityRecover : public Entity {
    public:
      EntityRecover() : Entity(EntityType::RECOVER) { }
      virtual const String name() { return "Recover"; }
    };

  }
}

#endif // HYPERTABLE_METALOGENTITY_H
