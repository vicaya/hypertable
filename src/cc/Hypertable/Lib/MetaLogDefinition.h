/** -*- c++ -*-
 * Copyright (C) 2010 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 2 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#ifndef HYPERTABLE_METALOGDEFINITION_H
#define HYPERTABLE_METALOGDEFINITION_H

#include "Common/ReferenceCount.h"
#include "Common/String.h"

#include "MetaLogEntity.h"

namespace Hypertable {
  namespace MetaLog {
    class Definition : public ReferenceCount {
    public:
      Definition(const char* backup_label) : m_backup_label(backup_label) { }
      virtual uint16_t version() = 0;
      virtual const char *name() = 0;
      virtual const char *backup_label() { return m_backup_label.c_str(); }
      virtual Entity *create(const EntityHeader &header) = 0;

    private:
      String m_backup_label;
    };
    typedef intrusive_ptr<Definition> DefinitionPtr;
  }

}

#endif // HYPERTABLE_METALOGDEFINITION_H
