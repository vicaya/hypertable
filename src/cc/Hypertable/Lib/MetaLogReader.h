/** -*- c++ -*-
 * Copyright (C) 2010 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 2 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#ifndef HYPERTABLE_METALOGREADER_H
#define HYPERTABLE_METALOGREADER_H

#include "Common/Filesystem.h"
#include "Common/ReferenceCount.h"

#include <map>
#include <vector>

#include "MetaLogDefinition.h"


namespace Hypertable {

  namespace MetaLog {

    class Reader : public ReferenceCount {

    public:
      Reader(FilesystemPtr &fs, DefinitionPtr &definition);
      Reader(FilesystemPtr &fs, DefinitionPtr &definition, const String &path);
      void get_entities(std::vector<EntityPtr> &entities);
      void get_all_entities(std::vector<EntityPtr> &entities);
      void reload();
      int32_t next_file_number() { return m_next_filenum; }
      bool load_file(const String &fname);

    private:

      bool verify_backup(int32_t file_num);
      void read_header(int fd);

      FilesystemPtr m_fs;
      MetaLog::DefinitionPtr m_definition;
      String m_path;
      int32_t m_next_filenum;
      std::vector<int32_t> m_file_nums;
      std::map<EntityHeader, EntityPtr> m_entity_map;
      std::vector<EntityPtr> m_entities;
      size_t m_cur_offset;
      String m_backup_path;
    };
    typedef intrusive_ptr<Reader> ReaderPtr;
  }
}

#endif // HYPERTABLE_METALOGREADER_H
