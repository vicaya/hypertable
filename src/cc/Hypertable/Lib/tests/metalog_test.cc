/** -*- c++ -*-
 * Copyright (C) 2010 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 2 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include "Common/Compat.h"
#include "Common/FileUtils.h"
#include "Common/Init.h"
#include "Common/InetAddr.h"
#include "Common/Random.h"
#include "Common/StringExt.h"
#include "Common/Serialization.h"
#include "DfsBroker/Lib/Client.h"
#include "AsyncComm/Comm.h"
#include "AsyncComm/ReactorFactory.h"
#include "AsyncComm/ConnectionManager.h"

#include <iostream>
#include <fstream>


#include "Hypertable/Lib/Config.h"

#include "Hypertable/Lib/MetaLogDefinition.h"
#include "Hypertable/Lib/MetaLogEntity.h"
#include "Hypertable/Lib/MetaLogReader.h"
#include "Hypertable/Lib/MetaLogWriter.h"

using namespace Hypertable;
using namespace Config;
using namespace std;

namespace Hypertable {
  namespace MetaLog {

    class EntityGeneric : public Entity {
    public:
      EntityGeneric(int t) : Entity(t), value(1) { m_name = String("GenericEntity") + t; }
      EntityGeneric(const EntityHeader &header_) : Entity(header_), value(0) { m_name = String("GenericEntity") + header_.type; }
      virtual const String name() { return m_name; }
      virtual size_t encoded_length() const { return 4; }
      virtual void encode(uint8_t **bufp) const {
        Serialization::encode_i32(bufp, value);
      }
      virtual void decode(const uint8_t **bufp, size_t *remainp) {
        value = Serialization::decode_i32(bufp, remainp);
      }
      virtual void display(ostream &os) { os << "value=" << value; }
      void increment() { value++; }

    private:
      String m_name;
      int32_t value;
    };

    class TestDefinition : public Definition {
    public:
      TestDefinition() : Definition("bar") {}
      virtual uint16_t version() { return 1; }
      virtual const char *name() { return "foo"; }
      virtual Entity *create(const EntityHeader &header) {
        return new EntityGeneric(header);
      }
    };

  }
}

namespace {

  struct MyPolicy : Policy {
    static void init_options() {
      cmdline_desc().add_options()
        ("save,s", "Don't delete generated the log files")
        ;
    }
  };

  typedef Meta::list<MyPolicy, DfsClientPolicy, DefaultCommPolicy> Policies;

  MetaLog::DefinitionPtr g_test_definition = new MetaLog::TestDefinition();

  vector<MetaLog::EntityPtr> g_entities;

  void create_entities(int count) {
    for (size_t i=65536; i<=(size_t)65536+count; i++)
      g_entities.push_back( new MetaLog::EntityGeneric(i) );
  }

  void randomly_change_states(MetaLog::WriterPtr &writer) {
    MetaLog::EntityGeneric *generic_entity;
    size_t j;
    for (size_t i=0; i<256; i++) {
      j = Random::number32() % g_entities.size();
      generic_entity = (MetaLog::EntityGeneric *)g_entities[j].get();
      if (generic_entity) {
        generic_entity->increment();
        if ((i%127)==0) {
          writer->record_removal(generic_entity);
          g_entities[j] = 0;
        }
        else
          writer->record_state(generic_entity);
      }
    }
  }

  void display_entities(ofstream &out) {
    for (size_t i=0; i<g_entities.size(); i++) {
      if (g_entities[i])
        out << *g_entities[i] << "\n";
    }
    out << flush;
  }

} // local namespace


int
main(int ac, char *av[]) {
  try {
    init_with_policies<Policies>(ac, av);

    int timeout = has("dfs-timeout") ? get_i32("dfs-timeout") : 180000;
    String host = get_str("dfs-host");
    uint16_t port = get_i16("dfs-port");

    DfsBroker::Client *client = new DfsBroker::Client(host, port, timeout);

    if (!client->wait_for_connection(timeout)) {
      HT_ERROR_OUT <<"Unable to connect to DFS: "<< host <<':'<< port << HT_END;
      return 1;
    }

    FilesystemPtr fs = client;
    String testdir = format("/metalog%09d", (int)getpid());
    MetaLog::WriterPtr writer;
    MetaLog::ReaderPtr reader;

    fs->mkdirs(testdir);

    MetaLog::EntityHeader::display_timestamp = false;

    /**
     *  Write initital log
     */

    writer = new MetaLog::Writer(fs, g_test_definition,
                                 testdir + "/" + g_test_definition->name(),
                                 g_entities);

    create_entities(32);

    {
      ofstream out("metalog_test.out");
      randomly_change_states(writer);
      writer = 0;
      reader = new MetaLog::Reader(fs, g_test_definition,
                                   testdir + "/" + g_test_definition->name());
      g_entities.clear();
      reader->get_entities(g_entities);
      display_entities(out);
      reader = 0;
      HT_ASSERT(FileUtils::size("metalog_test.out") == FileUtils::size("metalog_test.golden"));
    }

    /**
     *  Write some more
     */

    writer = new MetaLog::Writer(fs, g_test_definition,
                                 testdir + "/" + g_test_definition->name(),
                                 g_entities);

    {
      ofstream out("metalog_test2.out");
      randomly_change_states(writer);
      writer = 0;
      reader = new MetaLog::Reader(fs, g_test_definition,
                                   testdir + "/" + g_test_definition->name());
      g_entities.clear();
      reader->get_entities(g_entities);
      display_entities(out);
      reader = 0;
      HT_ASSERT(FileUtils::size("metalog_test2.out") == FileUtils::size("metalog_test2.golden"));
    }

    /**
     *  Write another log and skip the RECOVER entry
     */

    MetaLog::Writer::skip_recover_entry = true;

    writer = new MetaLog::Writer(fs, g_test_definition,
                                 testdir + "/" + g_test_definition->name(),
                                 g_entities);

    {
      ofstream out("metalog_test3.out");
      randomly_change_states(writer);
      writer = 0;
      reader = new MetaLog::Reader(fs, g_test_definition,
                                   testdir + "/" + g_test_definition->name());
      g_entities.clear();
      reader->get_entities(g_entities);
      display_entities(out);
      reader = 0;
      HT_ASSERT(FileUtils::size("metalog_test3.out") == FileUtils::size("metalog_test2.golden"));
    }

    if (!has("save"))
      fs->rmdir(testdir);
  }
  catch (Exception &e) {
    HT_ERROR_OUT << e << HT_END;
    return 1;
  }
  return 0;
}
