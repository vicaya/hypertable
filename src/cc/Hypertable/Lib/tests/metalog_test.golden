{MetaLog::Entity GenericEntity65536 header={type=65536,checksum=167773440,id=2,flags=0,length=4} payload={value=5}
{MetaLog::Entity GenericEntity65537 header={type=65537,checksum=369101568,id=3,flags=0,length=4} payload={value=11}
{MetaLog::Entity GenericEntity65538 header={type=65538,checksum=268437504,id=4,flags=0,length=4} payload={value=8}
{MetaLog::Entity GenericEntity65539 header={type=65539,checksum=335546880,id=5,flags=0,length=4} payload={value=10}
{MetaLog::Entity GenericEntity65540 header={type=65540,checksum=234882816,id=6,flags=0,length=4} payload={value=7}
{MetaLog::Entity GenericEntity65541 header={type=65541,checksum=301992192,id=7,flags=0,length=4} payload={value=9}
{MetaLog::Entity GenericEntity65543 header={type=65543,checksum=301992192,id=9,flags=0,length=4} payload={value=9}
{MetaLog::Entity GenericEntity65544 header={type=65544,checksum=234882816,id=10,flags=0,length=4} payload={value=7}
{MetaLog::Entity GenericEntity65545 header={type=65545,checksum=335546880,id=11,flags=0,length=4} payload={value=10}
{MetaLog::Entity GenericEntity65546 header={type=65546,checksum=503320320,id=12,flags=0,length=4} payload={value=15}
{MetaLog::Entity GenericEntity65547 header={type=65547,checksum=335546880,id=13,flags=0,length=4} payload={value=10}
{MetaLog::Entity GenericEntity65548 header={type=65548,checksum=234882816,id=14,flags=0,length=4} payload={value=7}
{MetaLog::Entity GenericEntity65549 header={type=65549,checksum=469765632,id=15,flags=0,length=4} payload={value=14}
{MetaLog::Entity GenericEntity65550 header={type=65550,checksum=369101568,id=16,flags=0,length=4} payload={value=11}
{MetaLog::Entity GenericEntity65551 header={type=65551,checksum=301992192,id=17,flags=0,length=4} payload={value=9}
{MetaLog::Entity GenericEntity65552 header={type=65552,checksum=301992192,id=18,flags=0,length=4} payload={value=9}
{MetaLog::Entity GenericEntity65554 header={type=65554,checksum=268437504,id=20,flags=0,length=4} payload={value=8}
{MetaLog::Entity GenericEntity65555 header={type=65555,checksum=335546880,id=21,flags=0,length=4} payload={value=10}
{MetaLog::Entity GenericEntity65556 header={type=65556,checksum=301992192,id=22,flags=0,length=4} payload={value=9}
{MetaLog::Entity GenericEntity65557 header={type=65557,checksum=369101568,id=23,flags=0,length=4} payload={value=11}
{MetaLog::Entity GenericEntity65558 header={type=65558,checksum=369101568,id=24,flags=0,length=4} payload={value=11}
{MetaLog::Entity GenericEntity65559 header={type=65559,checksum=268437504,id=25,flags=0,length=4} payload={value=8}
{MetaLog::Entity GenericEntity65560 header={type=65560,checksum=201328128,id=26,flags=0,length=4} payload={value=6}
{MetaLog::Entity GenericEntity65561 header={type=65561,checksum=201328128,id=27,flags=0,length=4} payload={value=6}
{MetaLog::Entity GenericEntity65562 header={type=65562,checksum=234882816,id=28,flags=0,length=4} payload={value=7}
{MetaLog::Entity GenericEntity65564 header={type=65564,checksum=167773440,id=30,flags=0,length=4} payload={value=5}
{MetaLog::Entity GenericEntity65565 header={type=65565,checksum=167773440,id=31,flags=0,length=4} payload={value=5}
{MetaLog::Entity GenericEntity65566 header={type=65566,checksum=234882816,id=32,flags=0,length=4} payload={value=7}
{MetaLog::Entity GenericEntity65567 header={type=65567,checksum=100664064,id=33,flags=0,length=4} payload={value=3}
{MetaLog::Entity GenericEntity65568 header={type=65568,checksum=335546880,id=34,flags=0,length=4} payload={value=10}
